### PR TITLE
gscan/gui: reduce impact on suites

### LINKED ISSUE #2262: gscan/gui: intelligent update frequency
Currently gscan performs a `scan_all` every time it updates. This is a slow process and so gscan is infrequently updated. This process is, however, only required to pick up new suites. Once gscan is aware of a suite it can query it directly which is much faster and thus can be performed more frequently.

### PR DESCRIPTION
**Pull request by matthewrmshin**
gscan:
* Separate time intervals for list+update (every ~5 minutes~ 1 minute) and update-only (every 15 seconds).
  * Both intervals are configurable.
* In a list+update, suite listing now done by walking `~/cylc-run/` in default mode, but will go back to `host:port` range scanning with the right options and arguments. E.g.:
  * `cylc gscan`: walk `~/cylc-run/`.
  * `cylc gscan --suite-owner=alice --suite-owner=bob`: walk `~alice/cylc-run/` and `~bob/cylc-run/`.
  * `cylc gscan -a`: scan port range for all known hosts, report suites of current user.
  * `cylc gscan -a --suite-owner=.*`: scan port range for all known hosts, report all suites.
  * `cylc gscan hostx hosty`: scan port range on `hostx` and `hosty`, report suites of current user.
  * And so on.
* In update-only, connect only to suites that were found by the latest listing.

gui:
* Consolidate 1 (initial connect) + 1 (poll update times) + 6 (data update) API calls into a single one.
  * (It will continue to require a separate API call when the graph view is switched on.)
* Server can use the previous client time (identified by client's UUID) to return data incrementally.
  * GUI updates its views only if server returns relevant values.
* Variable update frequency, taking into account time of last update and the mean duration of the last 10 main loops of the suite.
  * Do not update GUI faster than the suite's main loop.
  * Gradually reduce update frequency from 1 API call per second to 1 API call per 15 seconds (configurable) in quiet time.
  * Same logic and update frequency for stopped suite. It is no longer necessary to drop the update frequency of stopped suites like before - since cylc 7 the client should not create any network traffic when it is unable to load the content in the contact file.

Close #2262.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
(This is now ready for an initial look.)

**Comment by oliver-sanders** [comment]
Profiling results obtained from running the following suite:
```ini
[cylc]
    [[parameters]]
        task = 1..5
[scheduling]
    initial cycle point = 20000101T00Z
    final cycle point = 20000105T00Z
    [[dependencies]]
        [[[T00]]]
            graph = """
                task<task=5>[-P1D] => task<task=1>
                task<task-1> => task<task>
            """
[runtime]
    [[task<task>]]
        script = sleep 1
```

| branch | Metric | 1 x gcylc | 5 x gscan | No GUIs |
| --- | --- | --- | --- | --- |
| 7.5.0-71-g1e0cdb4c6 | CPU (total) | **7.54** | 4.46 |4.34 |
| | Memory | 29120 | 29024 | 29008 |
| matthewrmshin:cylc-scan-local | CPU (total) | **5.67** | 4.68 |4.60 |
| | Memory | 29204 | 29224 | 29060 |

**Comment by matthewrmshin** [comment]
gscan is perhaps quite different here. While everyone's gscan is no longer swamping everyone else's suites, I have tuned it to poll your own suites more frequently. So in the past, if we have 100 users, each running a single gscan, each running suite in the pool will be hit 100 times per minute. After the change, with the normal use case, each suite will only be hit 4 times per minute by the owner's gscan.

**Comment by hjoliver** [comment]
(minor conflicts)

**Comment by matthewrmshin** [comment]
Branch rebased.

File: lib/cylc/gui/updater.py [hunk 1]
```diff
@@ -471,83 +207,142 @@ def refilter(self):
...
+    def _update_err_log(self, my_state):
+        """Update suite err log if necessary."""
+        self.err_log_lines += my_state['err_content'].splitlines()
+        self.err_log_lines = self.err_log_lines[-10:]
+        gobject.idle_add(
+            self.info_bar.set_log, "\n".join(self.err_log_lines),
+            my_state['err_size'])
+
+    def _update_state_summary(self, my_state):
+        """Retrieve suite summary."""
```
Comment (hjoliver) [inline lib/cylc/gui/updater.py]: "Display suite summary." ?

File: lib/cylc/gui/scanutil.py [hunk 2]
```diff
@@ -501,41 +505,101 @@ def call_cylc_command(keys, command_id):
...
+        if owner_pattern is None:
+            # Run directory of current user only
+            run_dirs = [GLOBAL_CFG.get_host_item('run directory')]
+        else:
+            # Run directory of all users matching "owner_pattern".
+            # But skip those with /nologin or /false shells
+            run_dirs = []
+            skips = ('/false', '/nologin')
+            for pwent in getpwall():
+                if any(pwent.pw_shell.endswith(s) for s in (skip_shells)):
```
Comment (hjoliver) [inline lib/cylc/gui/scanutil.py]: `skip_shells` is not defined - presumably should be `skips`?

**Comment by hjoliver** [comment]
The description above is a bit inaccurate.  for `gscan` the option `--owner` should be `... --suite-owner`; and `-A` should be `-a`?

**Comment by hjoliver** [comment]
Default scan vs walk behaviour does not seem to be as advertised.  A plain `cylc gscan` (no args) is scanning (according to `--debug` output) and shows other users' suites.

**Comment by matthewrmshin** [comment]
(Numerous merge and re-base of this branch may have broken the intended logic. I'll take a quick look.)

File: lib/cylc/gui/scanutil.py [hunk 3]
```diff
@@ -501,41 +505,101 @@ def call_cylc_command(keys, command_id):
...
+                    continue
+                if owner_pattern.match(pwent.pw_name):
+                    run_dirs.append(GLOBAL_CFG.get_host_item(
+                        'run directory', owner=pwent.pw_name))
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/gui/scanutil.py]: OK. I have forgotten that `get_host_item` calls `os.path.expanduser`.

Looping over all pw entries appears to be very quick.

**Comment by matthewrmshin** [comment]
(Yes, definitely a bit broken, but I think I know why.)

**Comment by matthewrmshin** [comment]
@hjoliver Apart from the title bar/status bar suggestion. Everything should now be good.

**Comment by matthewrmshin** [comment]
(I'll add some debug stuff in the logic to help you test this out.)

**Comment by hjoliver** [comment]
(I'm on leave today - will get back to this next...)

**Comment by matthewrmshin** [comment]
I have increased the suite listing frequency to PT1M by default. This is probably not too bad - I think most usage of `gscan` would be the default user home directory scan, so should not cause any issues.

**Pull request merged**